### PR TITLE
[None][feat] Add Minimax H3 Parallel Tiled VAE Decoding and Ulysses Support

### PR DESCRIPTION
**Pull request by yibinl-nvidia**
## Dev Engineer Review

- Adds MiniMax-H3 Ulysses sequence parallelism with sequence padding, sharding, gathering, and mask propagation.
- Restricts multi-GPU execution to pure Ulysses with `VANILLA` attention.
- Adds configurable tiled VAE decoding with distributed tile execution, overlap blending, temporal chunking, and fallback paths.
- Preserves explicit VAE option values, including `False`, during pipeline configuration.
- Adds an 8-GPU BF16 example with parallel VAE tiling.
- Verify distributed collective coordination, tile assignment, output parity, and compatibility with existing pipeline configuration.
- Review finding counts and test execution results are unavailable.

## QA Engineer Review

- Added pipeline tests for VAE option propagation and Ulysses topology and attention validation.
- Added tiled VAE tests for distributed output parity, excess ranks, collective counts, fallback behavior, invalid configuration, and compression-ratio alignment.
- Added transformer tests for padding, masks, sharding, gathering, rotary embeddings, and baseline parity.
- Added both `test_minimax_h3_pipeline.py` and `test_minimax_h3_tiled_vae.py` to `tests/integration/test_lists/test-db/l0_cpu.yml`.
- Added a VAE benchmark for sequential, parallel, and untiled modes. It is not a CI or manual-QA test entry.
- Coverage verdict: **sufficient**. Test execution status is unavailable.

## Per-File QA Perspective

- `docs/source/models/visual-generation.md`: Documents MiniMax-H3 Ulysses support and unsupported configurations. Verify the documentation matches runtime validation.
- `examples/visual_gen/README.md`: Documents Ulysses and VAE tiling options and constraints. Verify the documented keys match configuration parsing.
- `examples/visual_gen/configs/minimax-h3-bf16-8gpu.yaml`: Adds an 8-GPU Ulysses and parallel VAE configuration. Verify it loads and runs on the documented setup.
- `tensorrt_llm/_torch/visual_gen/config.py`: Preserves VAE tiling options in pipeline configuration. Verify explicit `False` values remain effective.
- `tensorrt_llm/_torch/visual_gen/models/minimax_h3/pipeline_minimax_h3.py`: Enables tiled VAE loading and validates Ulysses settings. Verify invalid topology and attention settings fail correctly.
- `tensorrt_llm/_torch/visual_gen/models/minimax_h3/tiled_vae.py`: Adds sequential and distributed tiled decoding. Verify collectives, idle ranks, blending, parity, and fallback behavior.
- `tensorrt_llm/_torch/visual_gen/models/minimax_h3/transformer_minimax_h3.py`: Adds padded Ulysses sequence processing. Verify packed and pre-padded inputs across supported world sizes.
- `tests/integration/test_lists/test-db/l0_cpu.yml`: Adds the pipeline and tiled VAE tests to CPU CI. Both changed test files are listed.
- `tests/unittest/_torch/visual_gen/test_minimax_h3_pipeline.py`: Covers configuration propagation and Ulysses validation. It is listed in `l0_cpu.yml`.
- `tests/unittest/_torch/visual_gen/test_minimax_h3_tiled_vae.py`: Covers decoding parity, fallback behavior, validation, and collective usage. It is listed in `l0_cpu.yml`.
- `tests/unittest/_torch/visual_gen/test_minimax_h3_transformer.py`: Covers transformer and refiner sequence parallelism. No changed test-list entry for this file is shown.
- `tests/microbenchmarks/minimax_h3_vae.py`: Measures VAE latency, memory, output validity, and mode agreement. It is not listed as a CI or manual-QA test.

## Description

This PR adds MiniMax-H3 Ulysses sequence parallelism and parallel tiled VAE decoding.

## Performance

All results below use MiniMax-H3 on NVIDIA B200, BF16 transformer, VANILLA attention, 960×544 video, 124 frames, seed 42, and 27 denoising steps.

### End-to-end results: single-GPU baseline and 8×B200

| Configuration | End-to-end latency | Speedup vs earlier 1-GPU baseline |
|---|---:|---:|
| 1 B200, no Ulysses, sequential VAE tiling | 31.543 s | 1.00× |
| **8 B200, Ulysses size 8, sequential VAE tiling** | **9.419 s** | **3.35×** |
| **8 B200, Ulysses size 8, parallel VAE tiling** | **7.090 s** | **4.45×** |

## Correctness and test coverage

| Baseline | Candidate | Mean LPIPS | P95 LPIPS | Maximum LPIPS |
|---|---|---:|---:|---:|
| Single GPU, no Ulysses, sequential tiled VAE | 8-GPU Ulysses + parallel tiled VAE | **0.052** | **0.068** | **0.072** |
| 8-GPU Ulysses + sequential tiled VAE | 8-GPU Ulysses + parallel tiled VAE | **0** | **0** | **0** |

- **Exact output checks:** VAE-only decoded tensors have maximum absolute error = RMSE = **0**. 
- **111 selected tests passed:** 90 pipeline/VAE cases plus 21 targeted transformer Ulysses cases. The 35 new Ulysses cases cover supported/rejected configurations, token-refiner padding, global attention masks, rank-local modality/timestep indices and RoPE, existing padding, all-padding ranks, and output unpadding for 2/4/8 ranks. Rank transport is emulated in these focused unit tests; the eight-GPU E2E results above exercise real distributed execution. The pipeline/VAE tests are registered in `l0_cpu.yml`; transformer tests run in `l0_b200.yml`.

## Configuration and reproduction

Use the settings from `examples/visual_gen/configs/minimax-h3-bf16-4gpu.yaml`:

```yaml
parallel_config:
  ulysses_size: 8
pipeline_config:
  vae_use_tiling: true
  vae_tile_parallel: true
```

</details>

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [ ] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/visual_gen/models/minimax_h3/pipeline_minimax_h3.py [hunk 1]
```diff
@@ -146,7 +152,30 @@ def _request_generator(seed: int) -> torch.Generator:
 
     def __init__(self, pipeline_config: DiffusionPipelineConfig) -> None:
         if pipeline_config.mapping.world_size != 1:
-            raise NotImplementedError("MiniMax-H3 initial support is single-GPU only.")
+            vgm = pipeline_config.visual_gen_mapping
+            is_pure_ulysses = (
```
Comment (BowenFu) [inline tensorrt_llm/_torch/visual_gen/models/minimax_h3/pipeline_minimax_h3.py:156]: This is where H3 starts accepting multi-GPU configs, but as far as I can tell nothing in CI runs the new path with real collectives: the Ulysses tests use a fake `RankSharder` and identity blocks, and the tiled-VAE tests patch `torch.distributed.all_gather`. So the real all-to-all with the padded `key_padding_mask`, the refiner shard/gather, and the tile gather over `ulysses_group` never run together, and a group mix-up, mask/length mismatch, or collective hang would slip through.

Could you add a small 2-GPU test under `tests/unittest/_torch/visual_gen/multi_gpu/` (like `test_flux_ulysses.py` / `test_ltx2_parallel_vae.py`) comparing a tiny random-weight H3 forward at `ulysses_size=2` against single-GPU with a packed length not divisible by 2, plus parallel-vs-sequential tiled decode parity, and list it in `l0_dgx_b200.yml`? The benchmark's parity check should be reusable.

I think this is required for this PR, since it's the only coverage of the distributed behavior being added.

File: docs/source/models/visual-generation.md [hunk 2]
```diff
@@ -65,7 +65,7 @@ Models are auto-detected from the checkpoint directory. Diffusers-format models
 | **Wan 2.2** | Yes | Yes | Yes | Yes | Yes | Yes | Yes | Yes | Yes | Yes | Yes | Yes | Yes | Yes | No |
 | **FastWan 2.2** | Yes | Yes | No | No | No | No | No | No | Yes | Yes | Yes | No | No | No | No |
 | **LTX-2** | Yes | Yes | Yes | Yes | No | Yes | Yes | No | No | Yes | Yes | Yes | Yes | No | No |
-| **MiniMax-H3** | Yes | Yes | No | No | No | No | No | No | No | Yes | Yes | No | No | No | No |
+| **MiniMax-H3** | Yes | Yes | No | No | No | No | Yes | No | No | Yes | Yes | No | No | No | No |
```
Comment (brnguyen2) [inline docs/source/models/visual-generation.md]: This PR adds parallel VAE decode for MiniMax-H3, but the **Parallel VAE** column in this row still reads `No`. Flip it to `Yes` (a footnote that it is tile-parallel over the Ulysses group via `pipeline_config.vae_tile_parallel`, not `parallel_vae_size`, would help users who read the Wan/LTX-2 rows first).

File: tensorrt_llm/_torch/visual_gen/models/minimax_h3/pipeline_minimax_h3.py (hunk 1 above)
Comment (brnguyen2) [inline tensorrt_llm/_torch/visual_gen/models/minimax_h3/pipeline_minimax_h3.py]: `is_pure_ulysses` does not check `parallel_vae_size`. A user who follows the Wan/LTX-2 convention and sets `parallel_config.parallel_vae_size: 8` for H3 gets a warning from `BasePipeline.setup_parallel_vae` and single-rank decode, with no error. Add `and pipeline_config.parallel.parallel_vae_size == 1` here (with the error message pointing at `pipeline_config.vae_tile_parallel`), or map `parallel_vae_size == ulysses_size` onto tile parallelism.

File: tensorrt_llm/_torch/visual_gen/models/minimax_h3/tiled_vae.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+Temporal chunking, spatial overlap geometry, and blending remain owned by
+Diffusers. Only independent spatial decoder calls are distributed.
+"""
+
+import torch
+import torch.distributed as dist
+from diffusers import AutoencoderKLMiniMaxH3
+
+MINIMAX_H3_VAE_DEFAULTS = {
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/visual_gen/models/minimax_h3/tiled_vae.py]: `configure_tiling` now applies these values unconditionally, including on the existing 1-GPU path (`use_tiling=True`, 256/64). Please confirm in the PR description that the diffusers 0.40 `AutoencoderKLMiniMaxH3` already defaults to exactly this (tiling enabled, `tile_sample_min_height/width=256`, overlap 64). If any of those differ, single-GPU output changes and the LPIPS reference test in `l0_b200` would move. If they do match, consider reading the defaults from the loaded instance instead of hardcoding, so a diffusers bump cannot silently diverge.

File: tensorrt_llm/_torch/visual_gen/models/minimax_h3/transformer_minimax_h3.py [hunk 4]
```diff
@@ -343,9 +374,29 @@ def __init__(
         )
 
     def forward(self, hidden_states: torch.Tensor) -> torch.Tensor:
+        seq_len = hidden_states.shape[1]
+        pad = _padding_to_multiple(seq_len, self.sharder.size)
+        key_padding_mask = None
+        if pad:
+            valid = hidden_states.new_ones((hidden_states.shape[0], seq_len), dtype=torch.bool)
+            key_padding_mask = _pad_tensor_dim(valid, 1, pad, False)
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/visual_gen/models/minimax_h3/transformer_minimax_h3.py]: The refiner builds a `key_padding_mask` for Ulysses padding without the `_supports_key_padding_mask` guard the main `forward` has (line ~816). Today the pipeline's VANILLA-only gate prevents this from mattering, but the model is also constructed standalone (tests, benchmarks): with a non-VANILLA backend and `seq_len % ulysses_size != 0`, the mask is silently ignored and zero rows attend as real keys. Raise the same `NotImplementedError` here when `pad` is nonzero and the backend does not honour the mask.

File: tests/unittest/_torch/visual_gen/test_minimax_h3_tiled_vae.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # Tile-dependent values expose wrong ordering and overlap blending.
+        return (x + x.mean()).repeat_interleave(2, -2).repeat_interleave(2, -1)
+
+
+def _vae() -> TiledAutoencoderKLMiniMaxH3:
+    vae = TiledAutoencoderKLMiniMaxH3.__new__(TiledAutoencoderKLMiniMaxH3)
+    torch.nn.Module.__init__(vae)
+    vae.spatial_compression_ratio = 2
+    vae.post_quant_conv = torch.nn.Identity()
```
Comment (brnguyen2) [inline tests/unittest/_torch/visual_gen/test_minimax_h3_tiled_vae.py]: With `post_quant_conv = Identity`, the parity test cannot tell whether the override applies `post_quant_conv` per tile in the same way diffusers' `_decode_clip` does (e.g. before vs after slicing). A small random-weight `Conv3d(1, 1, 1)` stand-in would cover that axis at no cost.

File: tensorrt_llm/_torch/visual_gen/config.py [hunk 6]
```diff
@@ -604,6 +604,12 @@ def from_pretrained(
             if value:
                 extra_attrs[key] = value
 
+        # Preserve H3 VAE options, including explicit False values. Registry
+        # validation happens in PipelineLoader before this conversion.
+        for key in ("vae_use_tiling", "vae_tile_parallel", "vae_tile_size", "vae_tile_overlap"):
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/visual_gen/config.py]: Second per-model key list in a generic `from_pretrained` (LTX-2 above, H3 here), each with slightly different falsy handling. Since `PipelineLoader._resolve_pipeline_config` has already validated the dict against the registry `defaults`, copying the whole resolved dict into `extra_attrs` would remove both loops and the special-casing. Low priority, but the list will keep growing.

**Review by brnguyen2 — COMMENTED** [review]
Two whole-PR items, plus a tracking nit.

**Interaction with the existing parallel-VAE knob.** VisualGen already exposes `parallel_config.parallel_vae_size` (Wan, LTX-2) through `BasePipeline.setup_parallel_vae`. For H3 that path now logs a warning and no-ops (`pipeline.py:719`), while parallelism is enabled through the new `pipeline_config.vae_tile_parallel`. Keeping the H3 knob in `pipeline_config` is fine per ENGINEERING_CRITERIA §3 (the mechanism differs from the halo-exchange `ParallelVAEBase`), but the two must not conflict silently. Please either make the pure-Ulysses check in `[pipeline_minimax_h3.py:156](https://github.com/NVIDIA/TensorRT-LLM/pull/19506/files#diff-aec76ae708f58ddd1b47ff273b7d6ad92a02c2d59da0ad564af0b418c4480369R156)` also require `parallel_vae_size == 1` with an error pointing at `vae_tile_parallel`, or have `vae_tile_parallel` honour `parallel_vae_size` when it equals `ulysses_size`. Also document the relationship in the README section.

**No CI test runs a real collective.** `test_minimax_h3_tiled_vae.py` patches `all_gather`/`get_world_size`/`get_rank`, the transformer test replaces the sharder with a stub, and the e2e list has no Ulysses or tile-parallel H3 entry even though the description reports 8×B200 results. The benchmark's bitwise parity assert is the only real-collective check and is not in a test list. Please add an H3 entry to `tests/integration/defs/examples/visual_gen/test_visual_gen_multi_gpu.py` (the `ulysses4` Wan case is the precedent) covering `ulysses_size` with `vae_tile_parallel: true`, with an LPIPS-vs-1-GPU threshold.

**Tracking.** This is a nontrivial feature (new public knobs, new parallel path); the title carries `[None]`. Please attach a JIRA ticket.

Description hygiene: the description claims tile geometry "matches the sequential tiled decoder" and that defaults "use the checkpoint's spatial tiling". See the inline question on `MINIMAX_H3_VAE_DEFAULTS` about confirming those default values against the diffusers 0.40 class, since they are now applied on the single-GPU path too.
